pkg/cloudprovider/vsphere/nodemanager.go

```diff
@@ -198,29 +248,77 @@ func (nm *NodeManager) DiscoverNode(nodeID string, searchBy cm.FindVM) error {
 			continue
 		}
 
+		if (len(internalNetworkNameMap) > 0) && (len(externalNetworkNameMap) > 0) && v.Network == "" {
+			klog.Warningln("Skipping device because network name-based IP address detection is enabled and the \"Network\" field is not set on vNIC")
+			continue
+		}
+
 		// Only return a single IP address based on the preference of IPFamily
 		// Must break out of loop in the event of ipv6,ipv4 where the NIC does
 		// contain a valid IPv6 and IPV4 address
 		for _, family := range ipFamily {
 			ips := returnIPsFromSpecificFamily(family, v.IpAddress)
 
-			for _, ip := range ips {
-				klog.V(2).Infof("Adding IP: %s", ip)
+			if addressMatchingEnabled {
+				klog.V(2).Infof("Adding Hostname")
 				v1helper.AddToNodeAddresses(&addrs,
 					v1.NodeAddress{
-						Type:    v1.NodeExternalIP,
-						Address: ip,
-					}, v1.NodeAddress{
-						Type:    v1.NodeInternalIP,
-						Address: ip,
-					}, v1.NodeAddress{
 						Type:    v1.NodeHostName,
 						Address: oVM.Guest.HostName,
 					},
 				)
-
-				found = true
-				break
+				for _, ip := range ips {
+					parsedIP := net.ParseIP(ip)
+					if parsedIP == nil {
+						return err
+					}
+
+					internalNameMatched := isMatchingName(v.Network, internalNetworkNameMap)
+					internalSubnetMatched := isMatchingSubnet(parsedIP, externalNetworkSubnets)
+					if internalNameMatched || internalSubnetMatched {
+						klog.V(2).Infof("Adding Internal IP: %s", ip)
+						v1helper.AddToNodeAddresses(&addrs,
+							v1.NodeAddress{
+								Type:    v1.NodeInternalIP,
+								Address: ip,
+							},
+						)
+						found = true
+						break
```

Comment:
1. Should we or should we not `break` once we "classify" an ip address as Internal, without giving it a chance to be classified as External?
2.  Should we or should we not `break` once we actually find a matching ip address? Perhaps we should collect all available matches?

Here's how different implementations approach the *second* issue. You'll see links to both in-tree and CPI providers, since some CPI providers (notably, GCP and AWS) simply include legacy in-tree providers.

1. AWS. Uses all InternalIPs found: https://github.com/kubernetes/legacy-cloud-providers/blob/master/aws/aws.go#L1434-L1445
2. GCP. Uses the first InternalIP found and the first ExternalIP found: https://github.com/kubernetes/legacy-cloud-providers/blob/master/gce/gce_instances.go#L87-L94
3. Openstack. Uses all addresses found, classifying them based on config first: https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/cloudprovider/providers/openstack/openstack.go#L687-L711

Reply:
I think for now, we should stick to 1 InternalIP and  1 ExternalIP -- last time I checked kubelet always uses the first InternalIP it sees, which makes the behavior hard to predict since ordering is not always guaranteed. What do you think?

Reply:
I agree. Multiple IPs per type will needlessly complicate things. Until a concrete use case appears, ofc.